tools/make_patch.c

```diff
@@ -0,0 +1,437 @@
+		if (argc < 1) {
+			error_exit("Error: Invalid arguments for command: \"%s\"", command);
+		}
+		fprintf(output, command[strlen(command) - 1] == '_' ? "a%d: " : "a%d:", argc * 2);
+		for (int i = 0; i < argc; i++) {
+			int value = parse_arg_value(argv[i], false, symbols, current_hook->name);
+			if (value > 0xffff) {
+				error_exit("Error: Invalid value for \"%s\" argument: 0x%x", command, value);
+			}
+			if (i) {
+				putc(' ', output);
+			}
+			fprintf(output, isupper((unsigned char)command[0]) ? "%02X %02X": "%02x %02x", value & 0xff, value >> 8);
+		}
+
+	} else if (!strcmp(command, "db") || !strcmp(command, "DB") || !strcmp(command, "db_") || !strcmp(command, "DB_")) {
+		if (argc != 1) {
+			error_exit("Error: Invalid arguments for command: \"%s\"", command);
+		}
+		int value = parse_arg_value(argv[0], false, symbols, current_hook->name);
+		if (value > 0xff) {
+			error_exit("Error: Invalid value for \"%s\" argument: 0x%x", command, value);
+		}
+		fputs(command[strlen(command) - 1] == '_' ? "a1: " : "a1:", output);
+		fprintf(output, isupper((unsigned char)command[0]) ? "%02X" : "%02x", value);
+
+	} else if (!strcmp(command, "hex") || !strcmp(command, "HEX") || !strcmp(command, "HEx") || !strcmp(command, "Hex") || !strcmp(command, "heX") || !strcmp(command, "hEX")) {
+		if (argc != 1 && argc != 2) {
+			error_exit("Error: Invalid arguments for command: \"%s\"", command);
+		}
+		int value = parse_arg_value(argv[0], true, symbols, current_hook->name);
+		int padding = argc > 1 ? parse_number(argv[1], 0) : 2;
+		if (!strcmp(command, "HEx")) {
+			fprintf(output, "0x%0*X%02x", padding - 2, value >> 8, value & 0xff);
+		} else if (!strcmp(command, "Hex")) {
+			fprintf(output, "0x%0*X%03x", padding - 3, value >> 12, value & 0xfff);
+		} else if (!strcmp(command, "heX")) {
+			fprintf(output, "0x%0*x%02X", padding - 2, value >> 8, value & 0xff);
+		} else if (!strcmp(command, "hEX")) {
+			fprintf(output, "0x%0*x%03X", padding - 3, value >> 12, value & 0xfff);
+		} else {
+			fprintf(output, isupper((unsigned char)command[0]) ? "0x%0*X" : "0x%0*x", padding, value);
+		}
+
+	} else {
+		error_exit("Error: Unknown command: \"%s\"\n", command);
+	}
+}
+
+void skip_to_next_line(FILE *restrict input, FILE *restrict output) {
+	for (int c = getc(input); c != EOF; c = getc(input)) {
+		putc(c, output);
+		if (c == '\n' || c == '\r') {
+			break;
+		}
+	}
+}
+
+struct Buffer *process_template(const char *template_filename, const char *patch_filename, FILE *restrict new_rom, FILE *restrict orig_rom, const struct Symbol *symbols) {
+	FILE *input = xfopen(template_filename, 'r');
+	FILE *output = xfopen(patch_filename, 'w');
+	struct Buffer *buffer = buffer_create(1);
+
+	struct Buffer *patches = buffer_create(sizeof(struct Patch));
+
+	// The ROM checksum will always differ
+	buffer_append(patches, &(struct Patch){0x14e, 2});
+	// The Stadium data (see stadium.c) will always differ
+	unsigned int rom_size = (unsigned int)xfsize("", orig_rom);
```

Comment:
Is the cast truly necessary?

Reply:
`xfsize` returns a `long` which may be larger than an `unsigned int`, so I think so. Although I can only get a warning with clang's `-Wshorten-64-to-32`.

Reply:
You're only suppressing the warning with the explicit cast, not actually fixing the issue. Yeah you're losing bits if you have a >4GB file. Use `size_t` to preserve them.